src/relations/db.py

```diff
@@ -207,31 +204,40 @@ def set_up_relation(self, relation: Relation) -> bool:
                 )
             )
 
+            postgresql_version = None
+            try:
+                postgresql_version = self.charm.postgresql.get_postgresql_version()
+            except PostgreSQLGetPostgreSQLVersionError:
+                logger.exception(
+                    "Failed to retrieve the PostgreSQL version to initialise/update %s relation"
+                    % self.relation_name
+                )
+
             # Set the data in both application and unit data bag.
             # It's needed to run this logic on every relation changed event
             # setting the data again in the databag, otherwise the application charm that
             # is connecting to this database will receive a "database gone" event from the
             # old PostgreSQL library (ops-lib-pgsql) and the connection between the
             # application and this charm will not work.
-            for databag in [application_relation_databag, unit_relation_databag]:
-                updates = {
-                    "allowed-subnets": self._get_allowed_subnets(relation),
-                    "allowed-units": self._get_allowed_units(relation),
-                    "host": self.charm.endpoint,
-                    "master": primary,
-                    "port": DATABASE_PORT,
-                    "standbys": standbys,
-                    "version": self.charm.postgresql.get_postgresql_version(),
-                    "user": user,
-                    "password": password,
-                    "database": database,
-                    "extensions": ",".join(required_extensions),
-                }
-                databag.update(updates)
+            updates = {
+                "allowed-subnets": self._get_allowed_subnets(relation),
+                "allowed-units": self._get_allowed_units(relation),
+                "host": self.charm.endpoint,
+                "master": primary,
+                "port": DATABASE_PORT,
+                "standbys": standbys,
+                "user": user,
+                "password": password,
+                "database": database,
+                "extensions": ",".join(required_extensions),
+            }
+            if postgresql_version:
+                updates["version"] = postgresql_version
+            application_relation_databag.update(updates)
+            unit_relation_databag.update(updates)
         except (
             PostgreSQLCreateDatabaseError,
             PostgreSQLCreateUserError,
-            PostgreSQLGetPostgreSQLVersionError,
         ):
             self.charm.unit.status = BlockedStatus(
```

Comment:
We should most probably defer here. Should I change this as well?

Reply:
I think a `defer` would be good here. We just need to ensure that we can change the status to `Active` later when the database/user creation succeeds (which is what we are already missing today).